src/Core/src/Application.cs

```diff
@@ -0,0 +1,73 @@
+using System;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.Maui
+{
+	public abstract class Application : IApplication
+	{
+		static object? GlobalLock;
+
+		IServiceProvider? _serviceProvider;
+		IMauiContext? _context;
+
+		protected Application()
+		{
+			GlobalLock = new object();
+
+			if (Current != null)
+				throw new InvalidOperationException($"Only one {nameof(Application)} instance is allowed");
+
+			lock (GlobalLock)
+			{
+				Current = this;
+			}
+		}
+
+		static public Application? Current { get; internal set; }
+
+		public IServiceProvider? Services => _serviceProvider;
+
+		public IMauiContext? Context => _context;
+
+		public event EventHandler? Created;
+
+		public event EventHandler? Resumed;
+
+		public event EventHandler? Paused;
+
+		public event EventHandler? Stopped;
+
+		public abstract IWindow CreateWindow(IActivationState state);
+
+		public virtual void OnCreated()
+		{
+			Created?.Invoke(this, EventArgs.Empty);
+		}
+
+		public virtual void OnResumed()
+		{
+			Resumed?.Invoke(this, EventArgs.Empty);
+		}
+
+		public virtual void OnPaused()
+		{
+			Paused?.Invoke(this, EventArgs.Empty);
+		}
+
+		public virtual void OnStopped()
+		{
+			Stopped?.Invoke(this, EventArgs.Empty);
+		}
+
+		internal void SetServiceProvider(IServiceProvider provider)
+		{
+			_serviceProvider = provider;
+			SetHandlerContext(provider.GetService<IMauiContext>());
+		}
+
+		internal void SetHandlerContext(IMauiContext? context)
+		{
+			_context = context;
+		}
```

Comment:
I added a comment on the iOS app delegate about this should be pulled from the services, and passed in via the ctor using the DI. Not sure how far we want to take the DI, but if we do that, then we never need to _set_ things, we just ask the provider.

If someone doesn't want to use DI, then they don't have to and can either just use it for the ctor args as "the way" or not at all and never need to use it Because _we_ will use it to resolve handlers and whatnot, we can always get the `IApplication` or `IServiceProvider` when we have our ctors.